From 58e59c9f5c40ae4ab5a5f30e48faf2ca522ef34d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Tue, 23 Jan 2024 19:23:38 +0100
Subject: [PATCH] application: honor 'color-scheme' setting

The 'color-scheme' setting is the new system-wide dark preference introduced in GNOME 42. Set the dark theme based on this setting if it's not overridden by gnome-panel's own 'theme-variant' setting.
---
 gnome-panel/gp-application.c | 30 +++++++++++++++++++++++++++---
 1 file changed, 27 insertions(+), 3 deletions(-)

diff --git a/gnome-panel/gp-application.c b/gnome-panel/gp-application.c
index 75d8a32fc..4ff5fc598 100644
--- a/gnome-panel/gp-application.c
+++ b/gnome-panel/gp-application.c
@@ -28,6 +28,7 @@
 #include "panel-layout.h"
 #include "panel-lockdown.h"
 #include "panel-multiscreen.h"
+#include "panel-schemas.h"
 #include "panel-toplevel.h"
 #include "panel-widget.h"
 
@@ -44,6 +45,7 @@ struct _GpApplication
   GObject           parent;
 
   GSettings        *general_settings;
+  GSettings        *interface_settings;
 
   gulong            theme_name_id;
   gulong            prefer_dark_id;
@@ -189,13 +191,28 @@ theme_variant_changed_cb (GSettings     *settings,
                           GpApplication *self)
 {
   PanelThemeVariant variant;
+  char *color_scheme;
 
-  variant = g_settings_get_enum (settings, key);
+  variant = g_settings_get_enum (self->general_settings, PANEL_GENERAL_THEME_VARIANT_KEY);
 
   if (variant == PANEL_THEME_VARIANT_SYSTEM)
     {
-      gtk_settings_reset_property (gtk_settings_get_default (),
-                                   "gtk-application-prefer-dark-theme");
+      color_scheme = g_settings_get_string (self->interface_settings, "color-scheme");
+
+      if (g_strcmp0 (color_scheme, "prefer-dark") == 0)
+        {
+          g_object_set (gtk_settings_get_default (),
+                        "gtk-application-prefer-dark-theme",
+                        TRUE,
+                        NULL);
+        }
+      else
+        {
+          gtk_settings_reset_property (gtk_settings_get_default (),
+                                       "gtk-application-prefer-dark-theme");
+        }
+
+      g_free (color_scheme);
     }
   else
     {
@@ -344,6 +361,7 @@ gp_application_dispose (GObject *object)
   g_clear_pointer (&self->toplevels, g_hash_table_destroy);
 
   g_clear_object (&self->general_settings);
+  g_clear_object (&self->interface_settings);
   g_clear_object (&self->provider);
   g_clear_object (&self->module_manager);
   g_clear_object (&self->applet_manager);
@@ -368,6 +386,7 @@ static void
 gp_application_init (GpApplication *self)
 {
   self->general_settings = g_settings_new ("org.gnome.gnome-panel.general");
+  self->interface_settings = g_settings_new ("org.gnome.desktop.interface");
 
   g_signal_connect (self->general_settings,
                     "changed::theme-variant",
@@ -394,6 +413,11 @@ gp_application_init (GpApplication *self)
                     G_CALLBACK (panel_max_icon_size_cb),
                     self);
 
+  g_signal_connect (self->interface_settings,
+                    "changed::color-scheme",
+                    G_CALLBACK (theme_variant_changed_cb),
+                    self);
+
   self->theme_name_id = g_signal_connect (gtk_settings_get_default (),
                                           "notify::gtk-theme-name",
                                           G_CALLBACK (gtk_theme_name_changed_cb),
-- 
2.43.0

